include/nbl/builtin/hlsl/workgroup/shared_scan.hlsl

```diff
@@ -75,72 +73,59 @@ struct WorkgroupScanHead
             {
                 const uint prevLevelScan = sharedAccessor.main.get(gl_LocalInvocationIndex);
                 scan = subgroupOp(prevLevelScan);
-                if(isScan)
-                {
-                    sharedAccessor.main.set(scanStoreIndex, scan);
-                }
-            }
-            if(isScan)
-            {
-                scanStoreIndex += lastInvocationInLevel + 1u;
+                sharedAccessor.main.set(scanLoadIndex, scan);
             }
+            scanLoadIndex += lastInvocationInLevel + 1u;
         }
         if(lastInvocationInLevel >= glsl::gl_SubgroupSize())
         {
-            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
             if(participate)
             {
                 if(any(bool2(gl_LocalInvocationIndex == lastInvocationInLevel, isLastSubgroupInvocation))) {
-                    sharedAccessor.main.set(nextLevelStoreIndex, scan);
+                    sharedAccessor.main.set(scanLoadIndex - loadStoreIndexDiff, scan);
                 }
             }
             sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
             participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
             if(participate)
             {
-                const uint prevLevelScan = sharedAccessor.main.get(gl_LocalInvocationIndex);
+                const uint prevLevelScan = sharedAccessor.main.get(scanLoadIndex);
                 scan = subgroupOp(prevLevelScan);
-                if(isScan) {
-                    sharedAccessor.main.set(scanStoreIndex, scan);
-                }
             }
         }
-        return scan;
+        lastLevelScan = scan; // only invocations of gl_LocalInvocationIndex < gl_SubgroupSize will have correct values, rest will have garbage
     }
 };
 
-template<typename T, class Binop, class SharedAccessor>
+template<typename T, class Binop, class ScanHead, class SharedAccessor, bool isExclusive>
 struct WorkgroupScanTail
 {
-    bool isExclusive;
     T identity;
-    T firstLevelScan;
-    uint lastInvocation;
-    uint scanStoreIndex;
+    ScanHead head;
 
-    static WorkgroupScanTail create(bool isExclusive, T identity, T firstLevelScan, uint lastInvocation, uint scanStoreIndex)
+    static WorkgroupScanTail create(T identity, ScanHead head)
     {
-        WorkgroupScanTail<T, Binop, SharedAccessor> wst;
-        wst.isExclusive = isExclusive;
-        wst.identity = identity;
-        wst.firstLevelScan = firstLevelScan;
-        wst.lastInvocation = lastInvocation;
-        wst.scanStoreIndex = scanStoreIndex;
-        return wst;
+        WorkgroupScanTail<T, Binop, ScanHead, SharedAccessor, isExclusive> tail;
+        tail.identity = identity;
+        tail.head = head;
+        return tail;
     }
 
-    T operator()()
+    T operator()(NBL_REF_ARG(SharedAccessor) sharedAccessor)
     {
```

Comment:
https://github.com/Devsh-Graphics-Programming/Nabla/pull/519#discussion_r1327288972

call `head(sharedAccessor)` here, instead of outside

Reply:
or even better yet, don't hold the `head` as a member, just inherit from it!